pkg/proctree/proctree_feed.go

```diff
@@ -0,0 +1,257 @@
+package proctree
+
+import (
+	"github.com/aquasecurity/tracee/pkg/errfmt"
+	"github.com/aquasecurity/tracee/pkg/logger"
+	"github.com/aquasecurity/tracee/pkg/utils"
+)
+
+//
+// Feed process tree using signal events
+//
+
+type ForkFeed struct {
+	TimeStamp       uint64
+	ChildHash       uint32
+	ParentHash      uint32
+	LeaderHash      uint32
+	ParentTid       int32
+	ParentNsTid     int32
+	ParentPid       int32
+	ParentNsPid     int32
+	ParentStartTime uint64
+	LeaderTid       int32
+	LeaderNsTid     int32
+	LeaderPid       int32
+	LeaderNsPid     int32
+	LeaderStartTime uint64
+	ChildTid        int32
+	ChildNsTid      int32
+	ChildPid        int32
+	ChildNsPid      int32
+	ChildStartTime  uint64
+}
+
+// FeedFromFork feeds the process tree with a fork event.
+func (pt *ProcessTree) FeedFromFork(feed ForkFeed) error {
+	// DEBUG (TODO: remove this)
+	// file, _ := os.Open("/dev/null")
+	// file := os.Stdout
+	// fmt.Fprintf(file, "--\nFork event received:\n")
+	// fmt.Fprintf(file, "Parent Hash: %v\n", feed.ParentHash)
+	// fmt.Fprintf(file, "Leader Hash: %v\n", feed.LeaderHash)
+	// fmt.Fprintf(file, "Task   Hash: %v\n", feed.ChildHash)
+	// fmt.Fprintf(file, "PARENT:\ttid=%05d pid=%05d nspid=%05d nstid=%05d\n", feed.ParentTid, feed.ParentPid, feed.ParentNsPid, feed.ParentNsTid)
+	// fmt.Fprintf(file, "LEADER:\ttid=%05d pid=%05d nspid=%05d nstid=%05d\n", feed.LeaderTid, feed.LeaderPid, feed.LeaderNsPid, feed.LeaderNsTid)
+	// fmt.Fprintf(file, "CHILD: \ttid=%05d pid=%05d nspid=%05d nstid=%05d\n", feed.ChildTid, feed.ChildPid, feed.ChildNsPid, feed.ChildNsTid)
+	// END OF DEBUG
+
+	if feed.ChildHash == 0 || feed.ParentHash == 0 {
+		return errfmt.Errorf("invalid task hash")
+	}
+	if feed.ChildTid == 0 || feed.ChildPid == 0 {
+		return errfmt.Errorf("invalid child task")
+	}
+	if feed.ParentTid == 0 || feed.ParentPid == 0 {
+		return errfmt.Errorf("invalid parent task")
+	}
+
+	// Update the parent process (might already exist)
+
+	parent := pt.GetOrCreateProcessByHash(feed.ParentHash) // always a real process
+	parent.GetInfo().SetFeedAt(
+		TaskInfoFeed{
+			Tid:         int(feed.ParentTid),
+			Pid:         int(feed.ParentPid),
+			NsTid:       int(feed.ParentNsTid),
+			NsPid:       int(feed.ParentNsPid),
+			StartTimeNS: feed.ParentStartTime,
+			PPid:        -1, // do not change the parent PID
+			NsPPid:      -1, // do not change the ns parent PID
+			Uid:         -1, // do not change the parent UID
+			Gid:         -1, // do not change the parent GID
+		},
+		utils.NsSinceBootTimeToTime(feed.TimeStamp),
+	)
+	parent.AddChild(feed.LeaderHash) // add the leader as a child of the parent
```

Comment:
If this is not a child but a thread of the current process, do we still need to make this call?

Reply:
https://github.com/aquasecurity/tracee/pull/3364#discussion_r1320941337